cirq/ops/controlled_gate.py

```diff
@@ -12,52 +12,102 @@
+            control_qubits: The qubits that would act as controls.
+            num_control: Total number of control qubits.
         """
-        self.sub_gate = sub_gate
+        if control_qubits is None:
+            control_qubits = []
+        if num_controls < len(control_qubits):
+            raise ValueError('More specified control qubits than num_controls')
+
+        # Leave unspecified controls as Nones.
+        self.control_qubits = ([None]*(num_controls - # type: ignore
+                                       len(control_qubits)) +
+                               control_qubits)
+
+        # Flatten nested ControlledGates.
+        if isinstance(sub_gate, ControlledGate):
+            self.sub_gate = sub_gate.sub_gate # type: ignore
+            self.control_qubits += sub_gate.control_qubits
+        else:
+            self.sub_gate = sub_gate
+
+    def num_controls(self) -> int:
+        return len(self.control_qubits)
 
     def num_qubits(self) -> int:
-        return self.sub_gate.num_qubits() + 1
+        return self.sub_gate.num_qubits() + self.num_controls()
 
     def _decompose_(self, qubits):
         result = protocols.decompose_once_with_qubits(self.sub_gate,
-                                                      qubits[1:],
-                                                      NotImplemented)
+                                            qubits[self.num_controls():],
+                                            NotImplemented)
+
         if result is NotImplemented:
             return NotImplemented
 
-        return [cop.ControlledOperation(qubits[0], op) for op in result]
+        decomposed = []
+        for op in result:
+            decomposed.append(cop.ControlledOperation(
+                qubits[:self.num_controls()], op))
+        return decomposed
 
     def validate_args(self, qubits) -> None:
-        if len(qubits) < 1:
-            raise ValueError('No control qubit specified.')
-        self.sub_gate.validate_args(qubits[1:])
+        if len(qubits) < self.control_qubits.count(None):
+            raise ValueError('Not all control qubits specified.')
+        self.sub_gate.validate_args(qubits[self.control_qubits.count(None):])
+
+    def on(self, *qubits: raw_types.Qid) -> gate_operation.GateOperation:
+        if len(qubits) == 0:
+            raise ValueError(
+                "Applied a gate to an empty set of qubits. Gate: {!r}".format(
+                    self))
+        self.validate_args(qubits)
+
+        # Merge specified controls and new controls.
+        merged_controls = []
+        remaining_qubits = list(qubits)
+        for control in self.control_qubits:
+            if control is None:
+                merged_controls.append(remaining_qubits.pop(0))
+            else:
+                merged_controls.append(control)
+
+        # Convert to ControlledGate without control qubits that can be reused.
+        self.control_qubits = [None]*self.num_controls()
```

Comment:
Do not mutate `self`! If the user re-uses this gate instance, their controls will be gone.

Reply:
This was my intent. When you specify rest of the qubits and create an operation from a gate that has partial qubits prespecified, you should consume all the qubits together and the gate should become a regular gate that can be reused instead of being tied up to some of the qubits before. I can change this, if needed.

Reply:
Changed.